enhancements/kube-apiserver/api-webhook-supportability.md

```diff
@@ -0,0 +1,465 @@
+the affect resources. Notably, `controller_admission_duration_seconds` is not split by resource type
+in any way and hence just a cluster-global metric, and with that mostly useless.
+
+### Critical Alerts
+
+Based on the existing metrics from above, we will add the following critical alerts:
+
+- alert about latency per webhook `>1s` (`0.5s` for critical resources)
+- alert about failure count > 1% (both conversion and admission)
+
+We would like to have
+
+- alert about total webhook latency per resource >1s (0.5 for critical resources)
+
+but likely cannot implement that because of the lack of resource label (see section above). We have
+to discuss alternatives (new labels or new metric) upstream.
+
+Critical resources is a static set of resources that are critical to the operation of the cluster
+and likely bottlenecks for scalability and stability if admission is installed on them
+unconditionally (labelled with "performance" in the [Critical OCP API Resource
+spreadsheet](https://docs.google.com/spreadsheets/d/18A-bGQTcXrCB_sHAHBtzmsz0ma9fvFBtt5PY-VZ8T-Y)):
+
+- endpoints
+- events
+- pods
+- resourcequotas
+- apirequestcounts
+- endpointslices
+- clusterresourcequotas.
+
+### Degraded Operator
+
+The kube-apiserver operator will go degraded
+
+- if webhooks for virtual resources (SubjectAccessReview, ...) exist
+- if webhooks's service does not exist, or has no endpoints.
+
+Virtual resources are those that are not persisted in etcd (labelled as "virtual" in the
+[Critical OCP API Resource spreadsheet](https://docs.google.com/spreadsheets/d/18A-bGQTcXrCB_sHAHBtzmsz0ma9fvFBtt5PY-VZ8T-Y))):
+
+- bindings
+- `*`reviews.
+
+### Informational Alerts
+
+- webhook installed for core resources that are not whitelisted (via soft approval; restricted to
+  openshift-`*` namespaces)
+- webhook installed for security-sensitive resources  (TokenReviews, oauth tokens+clients) that are
+  not whitelisted (including non-openshift-`*` components).
+
+  This alert is not expressible with the metrics we have. We have to find a way either via
+  artificial metrics from the operator or by moving to some condition instead.
+
+Core resources are those shipped in a stock OCP cluster.
+
+Security sensitive resources (labelled with "security" in the [Critical OCP API Resource
+spreadsheet](https://docs.google.com/spreadsheets/d/18A-bGQTcXrCB_sHAHBtzmsz0ma9fvFBtt5PY-VZ8T-Y)):
+
+- secrets
+- serviceaccounts
+- mutatingwebhookconfigurations
+- validatingwebhookconfigurations
+- tokenreviews
+- certificatesigningrequests
+- credentialsrequests
+- oauthaccesstokens
+- oauthauthorizetokens
+- oauthclientauthorizations
+- oauthclients
+- useroauthaccesstokens
+- routes.
+
+### "Soft" Approval Process
+
+For RH-internal add-ons (those whose service lives in `openshift-*` namespaces) we add a
+```yaml
+webhook-approved.openshift.io: https://github.com/openshift/enhancement/pull/<pr>
```

Comment:
Will this get a validity check somehow or can anyone just add the annotation?

Reply:
like update. Everybody can fake it. That's why it is called "soft" approval. I have never seen any project to fake it in the community though.